### PR TITLE
Use EntityTypes

### PR DESCRIPTION
**Pull request by BillyGalbreath**

### CODE REVIEW

**Review by aikar — APPROVED** [review]
Overall maintenance burden is a concern but I think it might be ok. Mainly worried if files come in that are more 'broken'...

Hopefully mojang will do this themselves though. I'll ping them and see.

**Comment by BillyGalbreath** [comment]
@aikar There's a lot more, I only care for the EntityLiving ones, though ;)

**Review by zachbr — CHANGES_REQUESTED** [review]
Not sure how I feel about doing this for only a small portion of entities. It should probably be all or none.

**Comment by BillyGalbreath** [comment]
@zachbr I didn't do the others for the same maintainability concern @aikar brought up. This patch is actually for my Ridables plugin to work with less hacks.

**Comment by zachbr** [comment]
That's great and all, but not particularly relevant to Paper as a larger project.

**Comment by BillyGalbreath** [comment]
@zachbr Well, I mean it can be edited to include more when/if they are needed. I just didn't see a need/point in all the others at this time. Not like this would be the [first patch](https://github.com/PaperMC/Paper/blob/master/Spigot-API-Patches/0077-PreCreatureSpawnEvent.patch#L39-L40) with "more can be added later" submitted...

**Comment by zachbr** [comment]
Thats because you're still thinking of your own Ridables plugin. It's great that you want to clean up your own code and NMS should arguably be doing this anyway but that doesn't mean much to Paper. We shouldn't really be taking NMS changes just to make plugin based NMS hacks any easier but if we must I'd rather we do it consistently across the codebase rather than just doing it in the places only you care about.

**Comment by BillyGalbreath** [comment]
Then nevermind this. I don't have the time or need to include the others. You always say that we should hack NMS less and should send in PRs to prevent that. I tried. ¯\_(ツ)_/¯

**Comment by zachbr** [comment]
This is still hacking NMS, its just maintenance burden on us.
This response is pretty indicative of what I expected and all the more reason to reject this.

**Comment by zachbr** [comment]
Lots of internal discussion about this on Discord. Specifically there are concerns about edge cases in the non-living types as well as general maintainability. Outside of the edge cases I don't expect a ton of maintainability problems, but obviously we need to explore that to be sure.

I want to see how bad the edge cases around non-living entities are and we can go from there.

**Comment by zachbr** [comment]
After looking through the internals around this its quite clear that it would be a disaster to implement this for every single entity and then maintain it. There are simply too many edge cases.

**Review by zachbr — APPROVED** [review]

File: Spigot-Server-Patches/0386-Use-EntityTypes.patch [hunk 1]
```diff
@@ -0,0 +1,919 @@
+From 9b81b489a2d4237fbd36d6bfb6547bae03b79491 Mon Sep 17 00:00:00 2001
+From: BillyGalbreath <Blake.Galbreath@GMail.com>
+Date: Thu, 4 Oct 2018 10:08:02 -0500
+Subject: [PATCH] Use EntityTypes
```
Comment (kashike) [inline Spigot-Server-Patches/0386-Use-EntityTypes.patch]: ".. where possible"

**Review by kashike — CHANGES_REQUESTED** [review]

**Pull request merged**